pkg/auth/auth.go

```diff
@@ -147,24 +164,64 @@ func Login(ctx context.Context, systemContext *types.SystemContext, opts *LoginO
 		return nil
 	}
 	if unauthorized, ok := err.(docker.ErrUnauthorizedForCredentials); ok {
-		logrus.Debugf("error logging into %q: %v", server, unauthorized)
-		return errors.Errorf("error logging into %q: invalid username/password", server)
+		logrus.Debugf("error logging into %q: %v", key, unauthorized)
+		return errors.Errorf("error logging into %q: invalid username/password", key)
+	}
+	return errors.Wrapf(err, "authenticating creds for %q", key)
+}
+
+// getCredentialsForRef returns the auth config for the provided key as well as its registry.
+func getCredentialsForRef(systemContext *types.SystemContext, key string) (auth types.DockerAuthConfig, registry string, err error) {
+	trimmedKey := trimScheme(key)
+	if trimmedKey != key {
+		return auth, registry, errors.New("credentials key has https[s]:// prefix")
+	}
+
+	// Non-canonical references have to be handled by using the legacy API.
+	domain := getRegistryName(trimmedKey)
+	if trimmedKey == domain {
+		auth, err := config.GetCredentials(systemContext, domain)
+		return auth, domain, err
+	}
```

Comment:
I really dislike this code path. `reference.ParseNamed("quay.io")` fails because a "repository name must be canonical". We can't fulfill completely both use cases (pull and login/logout) with the same API (`GetCredentialsForRef`) and I don't see how we could solve it for now without using `GetCredentials`.

Reply:
Yeah, we don’t have a Go type for “a registry or a namespace or a repository”.

To an extent, this should be almost the only caller, so I’m fine with a _bit_ of awkwardness if it steers other callers towards doing the right thing.

Options:
- Remove the “deprecated” marking from `GetCredentials`. I am increasingly thinking that this might be the right thing to do — we have 3 or so legitimate users? OTOH it wouldn’t make this caller appreciably more elegant.
- Add a … `c/image/pkg/docker/config.GetCredentialsForStringKey` (?) that does ~what this `getCredentialsForRef` does. (Implement `GetCredentialsForRef` as a wrapper of `…ForStringKey`.) Add a scary warning that almost no-one should use it.
- Formalize the “{registry, namespace,repo}” scope concept in a c/image/docker/scopename subpackage, with a nice comprehensive parse/format/match/superset/… implementation, and start using it everywhere, including the `c/image/pkg/docker/config` API and possibly other parts of c/image that do scope matching. (And another sub package and/or a type for scopes that include domain wildcards?). Probably cleanest, gets this PR out of business of parsing the input in the `acceptRepositories` case at least, but expands scope dramatically.

Reply:
Yeah let's remove the deprecation for now: https://github.com/containers/image/pull/1305